### PR TITLE
o/hookstate/ctlcmd: add optional --pid and --apparmor-label arguments to "snapctl is-connected"

### PR DESCRIPTION
**Pull request by jhenstridge**
This PR is an implementation of @pedronis's `snapctl is-connected --pid` proposal described in this forum thread:

https://forum.snapcraft.io/t/autoconnect-requests-for-pulseaudio/18926/3?u=jamesh

Unlike regular `snapctl is-connected`, the `--pid` option only considers connections with the snap identified by the given process ID, as identified by `cgroup.SnapNameFromPid`.  The intention is to allow a snap daemon using local sockets check that the peer it is talking to has connected a given interface.

For example, a Pulse Audio snap may provide `audio-record` and `audio-playback` slots whose connection grant access to the daemon's socket.  The daemon can issue a `snapctl is-connected --pid $pid audio-record` to determine whether a particular client has been granted access to the microphone.

In addition to the normal 0/1 exit codes, the command may also return the following:
* 10: process ID does not refer to a snap
* 11: plug/slot is not connected, and process ID refers to a classic confined snap.

These both look like "false" to the shell, but can be used by services that want to communicate with host system processes or classic snaps.

There is a placeholder function to check whether to allow the feature to be used with a particular plug/slot.  At present, it hard codes it to allow use on slots of interface `cups-control`, `pulseaudio`, or `audio-record`.

### CODE REVIEW

**Comment by tillkamppeter** [comment]
CUPS is supposed to allow access from any non-snapped process and any classic-mode-snapped process in addition to standard-snapped processes plugging to "cups-control", so either the mentioned command should return "true" in all these cases or have special answers for non-snapped and classic-mode-snapped processes.

**Comment by tillkamppeter** [comment]
Is it also possible for CUPS to do this with an API call or only by calling this command line tool?

**Comment by jhenstridge** [comment]
> CUPS is supposed to allow access from any non-snapped process and any classic-mode-snapped process in addition to standard-snapped processes plugging to "cups-control"

Good point about classic confinement snaps.  I agree that for the Pulse Audio and CUPS use cases it would make sense to treat classic snaps like unconfined processes.  It does muddy things a bit though, since classic snaps can have plugs and slots.

> Is it also possible for CUPS to do this with an API call or only by calling this command line tool?

Yes.  The snapd-glib library includes a `snapd_client_run_snapctl_sync` method.  You'd probably need to call `snapd_client_set_socket_path` first to have the SnapdClient talk to `/run/snapd-snap.socket` instead of `/run/snapd.socket` though.

Since you'd be dealing with only a process ID, this would remove the need for determining and parsing the AppArmor label of the peer too.

File: overlord/hookstate/ctlcmd/is_connected.go
Comment (zyga) [inline overlord/hookstate/ctlcmd/is_connected.go:97]: I don't think so. We could have a tri-state answer: yes, no, not-a-snap

File: overlord/hookstate/ctlcmd/is_connected.go
Comment (zyga) [inline overlord/hookstate/ctlcmd/is_connected.go:40]: Not sure the description catches the essence. Not happy with my suggestion but what about:
```suggestion
	Pid int `long:"pid" description:"Process ID for a plausibly connected process"`
```

**Review by zyga — COMMENTED** [review]
Technically this looks correct and feels nice.

I left some comments. Please ping for a full review when not a draft.

File: overlord/hookstate/ctlcmd/is_connected.go
Comment (jhenstridge) [reply overlord/hookstate/ctlcmd/is_connected.go:97]: I agree.  I think distinct non-zero error codes for "not connected", "not connected, but pid is classic confined", and "not a snap" would be workable.

That way, callers that treat the exit status as a simple boolean will get the expected behaviour of determining whether a connection exists, while still allowing things like Pulse Audio or CUPS to service requests from the host system or classic snaps.

File: tests/main/snapctl-is-connected-pid/task.yaml
Comment (zyga) [inline tests/main/snapctl-is-connected-pid/task.yaml:16]: You can use `tests.cleanup prepare` at the top of `prepare`, then `tests.cleanup restore` in `restore`. This will let you run `tests.clenaup defer rm -f /snap` without repeating the condition in restore. Just do it here, it's nice for cleanup logic, as it's right next to the code that modifies the system.

File: tests/main/snapctl-is-connected-pid/task.yaml
Comment (zyga) [inline tests/main/snapctl-is-connected-pid/task.yaml:35]: This is unsafe. Use `set +e; "$@"; local retval=$?; set -e;`. Otherwise `set -e` is ignored for the entire duration of `fail`. Looking at the code below this is not a factor here, but it's really tricky to use right.

File: tests/main/snapctl-is-connected-pid/task.yaml
Comment (zyga) [inline tests/main/snapctl-is-connected-pid/task.yaml:66]: Some spacing is off here.

File: tests/main/snapctl-is-connected-pid/task.yaml
Comment (zyga) [inline tests/main/snapctl-is-connected-pid/task.yaml:32]: Since we are changing this, let's call it something more appropriate, `expect_exit_status` perhaps?

**Review by zyga — COMMENTED** [review]
Quick pass through the spread test (my favourite part of snapd patches)

**Comment by tillkamppeter** [comment]
@jhenstridge, anything still missing to drop the "Draft" status? I am waiting for this to get in, so that the CUPs Snap can finally get into the Snap Store.

**Comment by jhenstridge** [comment]
I've flipped the PR over to ready for review.  I added a check to restrict the feature to one of a few slot types.  That check probably should be changed a bit or moved to a more appropriate location, but I'm not sure where exactly.

**Comment by pedronis** [comment]
@jhenstridge I looked at this and spent some time thinking, let's chat about it

File: overlord/hookstate/ctlcmd/is_connected.go
Comment (pedronis) [inline overlord/hookstate/ctlcmd/is_connected.go:75]: all these interfaces may be deny-connection: true under some circumstances, which means sporting a slot doesn't mean the snap is intended/permitted to actually offer the interface.  Of course if the answer to "is connected" is true there's never problem disclosing that. As we discussed the issue is then whether we are worried about how much information we leak in the false cases,  the leakage is also different between the pid case and the apparmor label case, because in the latter case is very easy to establish if the other side is a snap at all or not without abusing this, the question then becomes about strict vs classic confinement.

Comment (alexmurray) [reply overlord/hookstate/ctlcmd/is_connected.go:75]: I am not sure what a malicious snap that provided an interface would be able to achieve with this in terms of information leakage - what kind of scenario are you envisaging here @pedronis? From what I can see here, this only allows a snap to query against it's own interfaces, so if a user has chosen to connect a snap to the malicious snap's interface, it should be assumed that the malicious snap can be aware of this. Can you elaborate?

**Comment by alexmurray** [comment]
This looks good to me from a security point of view, there are comprehensive tests and the implementation looks solid. Just waiting to follow up on Samuele's question above before signing off on this with security review.

File: overlord/hookstate/ctlcmd/is_connected.go
Comment (jhenstridge) [reply overlord/hookstate/ctlcmd/is_connected.go:75]: @alexmurray: with the pid based variant of this feature, it would let a snap enumerate the process ID space to check which ones belonged to snaps and which didn't.  That alone isn't a particularly useful, but is information that a snap wouldn't usually have and may enable some other attack we haven't thought of.

Comment (alexmurray) [reply overlord/hookstate/ctlcmd/is_connected.go:75]: The base apparmor profile already includes bits like:

```
  @{PROC}/@{pid}/ r,
...
  owner @{PROC}/@{pid}/cmdline r,
  owner @{PROC}/@{pid}/comm r,
```

This allows all snaps to enumerate processes running as their same user, and since it includes `cmdline` etc a snap can relatively easily identify other snap processes (running as the same user) by seeing if `/snap` is in `cmdline` - as such whilst we now allow snaps to identify other snaps regardless of the user, given most systems only have a very few number of users and most snaps will run as the primary user, I don't think this increases the information leakage in a significant manner. Nor am I aware of any way this enables any particular attack. 

One way to mitigate this would be to only permit this check if the specified `pid` has the same userid as the current `pid`, however, I suspect this would make this feature less useful to snaps like `pulseaudio` / `cups` etc which may run as root daemons etc.

So on balance, I am not concerned about this increased information leakage as a result.

**Comment by alexmurray** [comment]
Security review done as per the previous comments / discussion :+1:

File: overlord/hookstate/ctlcmd/is_connected.go
Comment (pedronis) [reply overlord/hookstate/ctlcmd/is_connected.go:75]: thanks for the input, I will mark this unblocked

File: overlord/hookstate/ctlcmd/is_connected.go
Comment (pedronis) [inline overlord/hookstate/ctlcmd/is_connected.go:62]: was there are a reason to pick 10 and 11, or it's mostly arbitrary?

File: overlord/hookstate/ctlcmd/is_connected.go
Comment (pedronis) [inline overlord/hookstate/ctlcmd/is_connected.go:121]: independently of the values we probably want to introduce constants:  

notASnapCode
classicSnapCode

**Review by pedronis — APPROVED** [review]
thanks, a couple of comments/questions

File: overlord/hookstate/ctlcmd/is_connected.go
Comment (pedronis) [inline overlord/hookstate/ctlcmd/is_connected.go:75]: we need to list these in the help? otherwise how can people know this limitation? or given we have agreed with security that there's no big worry of information leakage we shouldn't limit this? is still true that this doesn't make sense if there's no  peer/socket involved...

File: overlord/hookstate/ctlcmd/is_connected.go
Comment (jhenstridge) [reply overlord/hookstate/ctlcmd/is_connected.go:62]: It was an arbitrary choice: I just wanted to put some distance between them and the 0/1 exit statuses.  There is no code that depends on this interface yet, so we can easily change them to something else if you'd prefer.

File: overlord/hookstate/ctlcmd/is_connected.go
Comment (jhenstridge) [reply overlord/hookstate/ctlcmd/is_connected.go:75]: The main reason I would consider keeping this check is in case we discover some problem with the interface after merging.  It's much easier to change the interface if it is restricted to these fairly uncommon slot types.

I'd be happy to remove it though if you think it isn't needed.

Comment (pedronis) [reply overlord/hookstate/ctlcmd/is_connected.go:75]: I think it's fine to limit it initially as we learn about its usage, but I think these interfaces do need to be listed explicitly in the relevant part of the help for the command otherwise is just confusing.

File: overlord/hookstate/ctlcmd/is_connected.go
Comment (pedronis) [reply overlord/hookstate/ctlcmd/is_connected.go:62]: ok, it's kind of a nitpick but I would swap their use:

* 10 for a classic confined snap
* 11 if the process is not confined

**Review by pedronis — COMMENTED** [review]
@jhenstridge I answered to your last replies making some suggestions

File: overlord/hookstate/ctlcmd/is_connected_test.go
Comment (anonymouse64) [inline overlord/hookstate/ctlcmd/is_connected_test.go:80]:
```suggestion
	// snap1:plug1 does not use an allowed interface
```

File: overlord/hookstate/ctlcmd/is_connected_test.go
Comment (anonymouse64) [inline overlord/hookstate/ctlcmd/is_connected_test.go:84]:
```suggestion
	// snap1:slot1 does not use an allowed interface
```

File: overlord/hookstate/ctlcmd/is_connected_test.go
Comment (anonymouse64) [inline overlord/hookstate/ctlcmd/is_connected_test.go:108]:
```suggestion
	// snap1:plug1 does not use an allowed interface
```

File: overlord/hookstate/ctlcmd/is_connected_test.go
Comment (anonymouse64) [inline overlord/hookstate/ctlcmd/is_connected_test.go:112]:
```suggestion
	// snap1:slot1 does not use an allowed interface
```

**Review by anonymouse64 — APPROVED** [review]
a couple terminology changes to use more inclusive language, but other than that lgtm, thanks for working on this

**Pull request merged**
